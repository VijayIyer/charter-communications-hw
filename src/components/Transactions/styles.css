/* styles from https://www.w3schools.com/html/tryit.asp?filename=tryhtml_table_intro */
.transactions {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 5px;
}
.transactions__table {
  overflow-y: scroll;
  display: block;
  height: 200px;
  border-collapse: collapse;
  margin: 25px 0;
  border-radius: 5px 5px 0 0;
  font-size: 0.9em;
  font-family: sans-serif;
  min-width: 400px;
  box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);
}
.transactions__table thead tr {
  background-color: #009879;
  color: #ffffff;
  text-align: left;
}
.transactions__table th,
.transactions__table td {
  padding: 12px 15px;
}
.transactions__table tbody tr {
  border-bottom: 1px solid #dddddd;
}

.transactions__table tbody tr:nth-of-type(even) {
  background-color: #f3f3f3;
}

.transactions__table tbody tr:last-of-type {
  border-bottom: 2px solid #009879;
}

.transactions__regenerate-btn {
  background-color: var(--primary-bg-color);
  color: var(--primary-font-color);
  cursor: pointer;
  border: none;
  border-radius: 1em;
  padding: 1em;
}
.transactions__filters {
  display: flex;
  flex-wrap: wrap;
  gap: 5px;
  justify-content: center;
}
.transactions__filters * {
  border: none;
  margin: 2px;
  border-radius: 1em;
  padding: 1em;
}
.transactions__search-customer {
  position: relative;
  margin-bottom: 20px;
}
.transactions__search-customer__label {
  color: #009879;
  font-size: 13px;
  font-weight: normal;
  position: absolute;
  pointer-events: none;
  left: 15px;
  top: 11px;
  padding: 0 5px;
  background: #fff;
  transition: 0.2s ease all;
}
.transactions__search-customer__input {
  font-size: 12px;
  display: block;
  width: 100%;
  height: 36px;
  padding: 0 20px;
  background: #fff;
  color: #323840;
  border: 1px solid #009879;
  border-radius: 4px;
  box-sizing: border-box;
  &:focus {
    outline: none;
    ~ .transactions__search-customer__label {
      top: -8px;
      font-size: 13px;
    }
  }
}
.transactions__month-filter {
  background-color: var(--primary-bg-color);
  color: var(--primary-font-color);
}
