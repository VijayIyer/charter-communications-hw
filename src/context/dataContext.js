import { createContext, useEffect, useState } from "react";
import { fetchTransactionsData } from "../api/fetchTransactionsData";
export const dataContext = createContext({});
export const DataProvider = ({ children }) => {
  const [transactionsData, setTransactionsData] = useState([]);
  const [loadingTransactions, setLoadingTransactions] = useState(false);
  const [error, setError] = useState(false);
  const fetchData = () => {
    setLoadingTransactions(true);
    fetchTransactionsData()
      .then((data) => {
        setLoadingTransactions(false);
        setTransactionsData(data);
      })
      .catch((err) => {
        setError(true);
      });
  };
  useEffect(() => {
    fetchData();
  }, []);
  return (
    <dataContext.Provider
      value={{
        transactionsData,
        loadingTransactions,
        error,
        fetchData,
      }}
    >
      {children}
    </dataContext.Provider>
  );
};
