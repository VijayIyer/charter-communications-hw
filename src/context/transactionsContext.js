import { createContext, useEffect, useState } from "react";
import { fetchTransactionsData } from "../api/fetchTransactionsData";
export const transactionsContext = createContext({});
export const TransactionsProvider = ({ children }) => {
  const [transactionsData, setTransactionsData] = useState([]);
  const [loadingTransactions, setLoadingTransactions] = useState(false);
  useEffect(() => {
    setLoadingTransactions(true);
    fetchTransactionsData().then((data) => {
      setLoadingTransactions(false);
      setTransactionsData(data);
    });
  }, []);
  return (
    <transactionsContext.Provider
      value={{
        transactionsData,
        loadingTransactions,
      }}
    >
      {children}
    </transactionsContext.Provider>
  );
};
