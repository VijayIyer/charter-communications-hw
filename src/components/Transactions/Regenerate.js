import { useContext } from "react";
import { dataContext } from "../../context/dataContext";
export default function Regenerate() {
  const { fetchData } = useContext(dataContext);
  return (
    <button className='transactions__regenerate-btn' onClick={fetchData}>
      Regenerate Transactions Data
    </button>
  );
}
