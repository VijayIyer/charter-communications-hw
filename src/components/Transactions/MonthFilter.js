import { useState, useContext, useEffect } from "react";
import { dataContext } from "../../context/dataContext";
import { getUniqueElements, monthNames } from "../../utils";
import moment from "moment";
import { transactionsContext } from "../../context/transactionsContext";
export default function MonthFilter() {
  const [months, setMonths] = useState([]);
  const { monthFilter, setMonthFilter } = useContext(transactionsContext);
  const { transactionsData } = useContext(dataContext);
  useEffect(() => {
    if (transactionsData) {
      setMonths(
        getUniqueElements(
          transactionsData.map(
            (transaction) =>
              monthNames[moment(transaction.date).format("M") - 1]
          )
        )
      );
    }
  }, [transactionsData]);
  return (
    <div>
      <label>Month</label>
      <select
        className='transactions__month-filter'
        onChange={(e) => setMonthFilter(e.target.value)}
        value={monthFilter}
      >
        <option value=''>--Select a month ---</option>
        {months.map((month) => (
          <option key={month} value={month}>
            {month}
          </option>
        ))}
      </select>
    </div>
  );
}
