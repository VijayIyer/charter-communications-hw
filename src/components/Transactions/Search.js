import { useContext } from "react";
import { transactionsContext } from "../../context/transactionsContext";

export default function Search() {
  const { customerFilter, setCustomerFilter } = useContext(transactionsContext);
  return (
    <div className='transactions__search-customer'>
      <label className='transactions__search-customer__label'>Customer</label>
      <input
        type='text'
        className='transactions__search-customer__input'
        value={customerFilter}
        placeholder=' '
        onChange={(e) => setCustomerFilter(e.target.value)}
      />
    </div>
  );
}
