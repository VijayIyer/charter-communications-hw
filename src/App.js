import Transactions from "./components/Transactions";
import RewardsTable from "./components/Rewards/RewardsTable";
import { TransactionsProvider } from "./context/transactionsContext";

function App() {
  return (
    <div className='App'>
      <TransactionsProvider>
        <Transactions />
        <RewardsTable />
      </TransactionsProvider>
    </div>
  );
}

export default App;
