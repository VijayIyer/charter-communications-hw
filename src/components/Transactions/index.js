import TransactionsTable from "./TransactionsTable";
import Transaction from "./Transaction";
import MonthFilter from "./MonthFilter";
import Search from "./Search";
import Regenerate from "./Regenerate";
import "./styles.css";
export default function Transactions() {
  return (
    <div className='transactions'>
      <h3>Transactions:</h3>
      <Regenerate />
      <div className='transactions__filters'>
        <MonthFilter />
        <Search />
      </div>
      <TransactionsTable />
    </div>
  );
}
