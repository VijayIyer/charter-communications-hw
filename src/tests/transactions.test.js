import Transactions from "../components/Transactions";
import React from "react";
import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react";
import App from "../App";
import Regenerate from "../components/Transactions/Regenerate";
import Search from "../components/Transactions/Search";
import { transactionsContext } from "../context/transactionsContext";
import { dataContext } from "../context/dataContext";
import MonthFilter from "../components/Transactions/MonthFilter";

it("renders App correctly", () => {
  const { container } = render(<App />);
  expect(screen.getByText("Transactions:")).toBeInTheDocument();
  expect(screen.getByText("Show Rewards")).toBeInTheDocument();
});
describe(`tests snapshot`, () => {
  it(`should match snapshot`, () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
describe(`tests Regenerate component`, () => {
  it(`should contain the text 'Regenerate Transactions Data'`, () => {
    const { container } = render(<Regenerate />);
    expect(
      screen.getByText("Regenerate Transactions Data")
    ).toBeInTheDocument();
  });
  it(`calls fetchData function when 'Regenerate Transactions Data' button is clicked`, () => {
    const fetchData = jest.fn();
    const { container } = render(
      <dataContext.Provider value={{ fetchData }}>
        <transactionsContext.Provider value={null}>
          <Regenerate />
        </transactionsContext.Provider>
      </dataContext.Provider>
    );
    const regenerateComponent = container.getElementsByClassName(
      "transactions__regenerate-btn"
    )[0];
    fireEvent.click(regenerateComponent);
    expect(fetchData).toBeCalledTimes(1);
  });
});
describe(`tests Search component`, () => {
  it(`should contain label element with the text 'Customer:'`, () => {
    const { container } = render(
      <transactionsContext.Provider
        value={{ customerFilter: null, setCustomerFilter: null }}
      >
        <Search />
      </transactionsContext.Provider>
    );
    expect(screen.getByText("Customer:")).toBeInTheDocument();
    expect(container.getElementsByTagName("label").length).toEqual(1);
  });
  it(`should contain an element with class name 'transactions__search-customer'`, () => {
    const { container } = render(
      <transactionsContext.Provider
        value={{ customerFilter: null, setCustomerFilter: null }}
      >
        <Search />
      </transactionsContext.Provider>
    );
    expect(
      container.getElementsByClassName("transactions__search-customer").length
    ).toEqual(1);
  });
});
describe(`tests MonthFilter component`, () => {
  it(`should contain label with text 'See Transactions in Month:' and a select element`, () => {
    const { container } = render(
      <transactionsContext.Provider
        value={{ monthFilter: null, setMonthFilter: null }}
      >
        <MonthFilter />
      </transactionsContext.Provider>
    );
    expect(screen.getByText("See Transactions in Month:")).toBeInTheDocument();
    expect(container.getElementsByTagName("label").length).toEqual(1);
    expect(container.getElementsByTagName("select").length).toEqual(1);
  });
});
