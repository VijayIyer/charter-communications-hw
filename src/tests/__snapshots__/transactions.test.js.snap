// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests snapshot should match snapshot 1`] = `
<div
  className="App"
>
  <div
    className="transactions"
  >
    <h3>
      Transactions:
    </h3>
    <button
      className="transactions__regenerate-btn"
      onClick={[Function]}
    >
      Regenerate Transactions Data
    </button>
    <div
      className="transactions__filters"
    >
      <div>
        <label>
          See Transactions in Month:
        </label>
        <select
          className="transactions__month-filter"
          onChange={[Function]}
          value=""
        >
          <option
            value=""
          >
            --Select a month ---
          </option>
        </select>
      </div>
      <div>
        <label>
          Customer:
        </label>
        <input
          className="transactions__search-customer"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="transactions__table-wrapper"
    >
      <table
        className="transactions__table"
      >
        <thead>
          <tr>
            <th />
            <th>
              Customer Name
            </th>
            <th>
              Date
            </th>
            <th>
              Amount
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>
  <div
    className="rewards"
  >
    <button
      className="rewards__show-btn"
      onClick={[Function]}
    >
      Show Rewards
    </button>
  </div>
</div>
`;
